RegistrationApp/src/test/java/app/util/UserUtilsTest.java

```diff
@@ -0,0 +1,233 @@
+        users.add(createUser(112,"username1"+112, "SFirstName1"+112, "Last Name1"+112, "someone1@gmail.com2", LocalDate.of(1900,11,10),false,"+380502456781"));
+        List<User> filterUsers=utils.showAvailable(users);
+        assertEquals(8,users.size());
+        assertEquals(6,filterUsers.size());
+        assertIterableEquals(createUsers(),filterUsers);
+
+    }
+
+    @Test
+    void showOnlyWithPhone() {
+        UserUtils utils=new UserUtils();
+        List<User> users=createUsers();
+        users.add(createUser(111,"username1", "SFirstName1", "Last Name1", "someone1@gmail.com1", LocalDate.of(1900,11,10),false,null));
+        users.add(createUser(112,"username1"+112, "SFirstName1"+112, "Last Name1"+112, "someone1@gmail.com2", LocalDate.of(1900,11,10),false,""));
+        List<User> filterUsers=utils.showOnlyWithPhone(users);
+        assertEquals(8,users.size());
+        assertEquals(6,filterUsers.size());
+        assertIterableEquals(createUsers(),filterUsers);
+    }
+
+    @Test
+    void groupUsers() {
+        UserUtils utils=new UserUtils();
+        List<User> users=createUsers();
+        users.add(createUser(111,"username1"+1, "SFirstName1"+1, "ALast Name1"+1, "someone1@gmail.com1", LocalDate.of(1900,01,10),true,"+380502456781"));
+        users.add(createUser(112,"username1"+112, "SFirstName1"+112, "Last Name1"+112, "someone1@gmail.com2", LocalDate.of(1900,02,10),true,"+380502456781"));
+        Map<String, List<User>> groupUsers = utils.groupUsers(users);
+        assertEquals(8,users.size());
+        assertThat(groupUsers.size(),is(8));
+
+    }
+
+    @Test
+    void findByPattern() {
+        UserUtils utils=new UserUtils();
+        List<User> users=createUsers();
+        users.add(createUser(111,"username1", "SFirstName1", "ALast Name1", "someone1@gmail.com1", LocalDate.of(1900,01,10),true,"+380502456781"));
+        users.add(createUser(112,"username", "SFirstName1", "Last Name1", "someone1@gmail.com2", LocalDate.of(1900,02,10),true,"+380502456781"));
+        List<User> findUsers=utils.findByPattern(users,"[a-zA-Z]{3,}[0-9]{2,}");
+        assertEquals(6,findUsers.size());
+    }
+
+    @Test
+    void findEarliestDate() {
+        UserUtils utils=new UserUtils();
+        List<User> users=createUsers();
+        users.add(createUser(111,"username1"+0, "SFirstName1"+0, "Last Name1"+0+"", "someone1@gmail.com1", LocalDate.of(1900,11,10),true,"+380502456781"));
+        users.add(createUser(112,"username1"+112, "SFirstName1"+112, "Last Name1"+112, "someone1@gmail.com2", LocalDate.of(1901,11,10),true,"+380502456781"));
+        Map<Integer, LocalDate> groupedUsers = utils.findEarliestDate(users);
+        assertThat(groupedUsers.size(),is(1));
+        assertThat(groupedUsers, IsMapContaining.hasEntry(111, LocalDate.of(1900,11,10)));
+    }
+
+    @Test
+    void filterByYearSortByMonth() {
```

Comment:
you don't need to clear it

Reply:
here is the result with removed 'testUser.clear();'
![image](https://user-images.githubusercontent.com/117306917/222395120-339cb4d9-f14f-4fc1-acc0-dd2dc29a9cbb.png)

Reply:
you need to rewrite it, logic is unclear
why you create List<User> testUser = new ArrayList<>(); adding user there and test it? after that you clear it. Let's do it together on the lesson